lib/spack/llnl/util/filesystem.py

```diff
@@ -1712,66 +1715,140 @@ def find(root, files, recursive=True):
+    regexes = [re.compile(fnmatch.translate(os.path.normcase(x))) for x in globs]
 
-    # Make the path absolute to have absolute path returned
+    # Note later calls to os.scandir etc. return abspaths if the
+    # input is absolute, see https://docs.python.org/3/library/os.html#os.DirEntry.path
     root = os.path.abspath(root)
 
-    for search_file in search_files:
-        matches = glob.glob(os.path.join(root, search_file))
-        matches = [os.path.join(root, x) for x in matches]
-        found_files[search_file].extend(matches)
+    found_files = collections.defaultdict(list)
 
-    answer = []
-    for search_file in search_files:
-        answer.extend(found_files[search_file])
+    def _dir_id(stat_info):
+        # Note: on windows, st_ino is the file index and st_dev
+        # is the volume serial number. See
+        # https://github.com/python/cpython/blob/3.9/Python/fileutils.c
+        return (stat_info.st_ino, stat_info.st_dev)
+
+    def _log_file_access_issue(e):
+        errno_name = errno.errorcode.get(e.errno, "UNKNOWN")
+        tty.debug(f"find must skip {dir_entry.path}: {errno_name} {str(e)}")
+
+    visited_dirs = set([_dir_id(stat_root)])
+
+    # Each queue item stores the depth and path
+    # BFS traversal is important to ensure repeatable results given
+    # that we allow following symlinks to sibling dirs that are not
+    # as deep: we return the shortest path to a given file
+    dir_queue = collections.deque([(0, root)])
+    while dir_queue:
+        depth, next_dir = dir_queue.pop()
+        try:
+            dir_iter = os.scandir(next_dir)
+        except OSError:
+            # Most commonly, this would be a permissions issue, for
+            # example if we are scanning an external directory like /usr
+            continue
 
-    return answer
+        with dir_iter:
+            for dir_entry in dir_iter:
+                try:
+                    it_is_a_dir = dir_entry.is_dir(follow_symlinks=True)
+                except OSError as e:
+                    # Possible permission issue, or a symlink that cannot
+                    # be resolved (ELOOP).
+                    _log_file_access_issue(e)
+                    continue
+
+                if it_is_a_dir and (depth < max_depth):
+                    try:
+                        # The stat should be performed in a try/except block.
+                        # We repeat that here vs. moving to the above block
+                        # because we only want to call `stat` if we are below
+                        # our max_depth
+                        if sys.platform == "win32":
+                            # Note: st_ino/st_dev on DirEntry.stat are not set on
+                            # Windows, so we have to call os.stat
+                            stat_info = os.stat(dir_entry.path, follow_symlinks=True)
+                        else:
+                            stat_info = dir_entry.stat(follow_symlinks=True)
```

Comment:
> On non-Windows where DirEntry.stat(...) is actually usable, use it, since that avoids the second stat call for the symlinked dir case as you're using its cache from the previous is_dir call (could be minor in practice, but trivial to fix).

I have done so, but FYI, the documentation for `DirEntry.stat` claims that it always requires a system call: https://docs.python.org/3/library/os.html#os.DirEntry.stat

Reply:
point is it's cached from the previous `is_dir` call

Reply:
The docs claim it always requires a system call, and does not make an exception for caching (this isn't a problem with me, but I'm curious what you are basing this assertion off of).